src/process_utility.c

```diff
@@ -1209,12 +1238,122 @@ process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
 	ts_chunk_index_create_from_stmt(stmt, chunk->fd.id, chunk_relid, ht->fd.id, info->obj.objectId);
 }
 
+static void
+process_index_chunk_multitransaction(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+	CatalogSecurityContext sec_ctx;
+	Chunk *chunk;
+	Relation hypertable_rel;
+	Relation hypertable_index_rel;
+	Relation chunk_rel;
+
+	Assert(info->extended_options.multitransaction);
+
+	/*
+	 * commit the previous transaction (foreach_chunk requires a transcation
+	 * running, so we cannot commit before this point)
+	 */
+	if (IsTransactionOrTransactionBlock())
+		CommitTransactionCommand();
+
+	/* Start a new transaction for each relation. */
+	StartTransactionCommand();
+	MemoryContextSwitchTo(info->mctx);
+
+#ifdef DEBUG
+	if (info->extended_options.max_chunks == 0)
+		return;
+
+	if (info->extended_options.max_chunks > 0)
+		info->extended_options.max_chunks -= 1;
+
+	if (OidIsValid(info->extended_options.barrier_table))
+	{
+		/*
+		 * For insolation tests, and debugging, it's useful to be able to
+		 * pause CREATE INDEX immediately before it starts working on chunks.
+		 * We acquire and immediately release a lock on a barrier table to do
+		 * this.
+		 */
+		Relation barrier = relation_open(info->extended_options.barrier_table, AccessExclusiveLock);
+
+		relation_close(barrier, AccessExclusiveLock);
+	}
+#endif
+
+	/*
+	 * Change user since chunks are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In the single-tranaction
+	 * case, we do this once for the entire table.
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index, and the chunk to prevent
+	 * from being altered. Since we use the same relids across transactions,
+	 * there is a potential issue if the id gets reassigned between one
+	 * sub-transaction and the next. CLUSTER has a similar issue.
+	 *
+	 * We grab a ShareLock on the chunk, because that's what CREATE INDEX
+	 * does. For the hypertable, and the hypertable's index, we are ok using
+	 * the weaker AccessShareLock, since we only need to prevent the index
+	 * iteself from being ALTERed or DROPed during this part of index
+	 * creation. TODO: is the above correct?
+	 */
+	hypertable_rel = relation_open(ht->main_table_relid, AccessShareLock);
+	chunk_rel = relation_open(chunk_relid, ShareLock);
+	hypertable_index_rel = relation_open(info->obj.objectId, AccessShareLock);
```

Comment:
isn't there a session-level lock on this?

Reply:
From the docs on `SessionLock` you're supposed to take a regular lock on the relation anyway